src/System.Private.CoreLib/src/System/Threading/ThreadPool.Unix.cs

```diff
@@ -4,18 +4,279 @@
+
+
+        /// <summary>
+        /// The <see cref="WaitHandle"/> that was registered.
+        /// </summary>
+        internal WaitHandle Handle { get; }
+
+        /// <summary>
+        /// The time this handle times out at in ms.
+        /// </summary>
+        internal int TimeoutTimeMs { get; private set; }
+
+        private int TimeoutDurationMs { get; }
+
+        internal bool IsInfiniteTimeout => TimeoutDurationMs == -1;
+
+        internal void RestartTimeout(int currentTimeMs)
+        {
+            TimeoutTimeMs = currentTimeMs + TimeoutDurationMs;
+        }
+
+        /// <summary>
+        /// Whether or not the wait is a repeating wait.
+        /// </summary>
+        internal bool Repeating { get; }
+
+        /// <summary>
+        /// The <see cref="WaitHandle"/> the user passed in via <see cref="Unregister(WaitHandle)"/>.
+        /// </summary>
+        private SafeWaitHandle UserUnregisterWaitHandle { get; set; } = new SafeWaitHandle((IntPtr)(-1), false); // Initialize with an invalid handle like CoreCLR
+
+        private IntPtr UserUnregisterWaitHandleValue { get; set; } = new IntPtr(-1);
+
+        internal bool IsBlocking => UserUnregisterWaitHandleValue == (IntPtr)(-1);
+
+        /// <summary>
+        /// The <see cref="ClrThreadPool.WaitThread"/> this <see cref="RegisteredWaitHandle"/> was registered on.
+        /// </summary>
+        internal ClrThreadPool.WaitThread WaitThread { get; set; }
+
+        /// <summary>
+        /// The number of callbacks that are currently queued on the Thread Pool or executing.
+        /// </summary>
+        private int _numRequestedCallbacks;
+
+        private LowLevelLock _callbackLock = new LowLevelLock();
+
+        /// <summary>
+        /// Notes if we need to signal the user's unregister event after all callbacks complete.
+        /// </summary>
+        private bool _signalAfterCallbacksComplete;
+
+        private int _unregisterCalled;
+
+        private bool _unregistered;
+
+        private AutoResetEvent _callbacksComplete;
+
+        private AutoResetEvent _removed;
+
+        /// <summary>
+        /// Unregisters this wait handle registration from the wait threads.
+        /// </summary>
+        /// <param name="waitObject">The event to signal when the handle is unregistered.</param>
+        /// <returns>If the handle was successfully marked to be removed and the provided wait handle was set as the user provided event.</returns>
+        /// <remarks>
+        /// This method will only return true on the first call.
+        /// Passing in a wait handle with a value of -1 will result in a blocking wait, where Unregister will not return until the full unregistration is completed.
+        /// </remarks>
         public bool Unregister(WaitHandle waitObject)
         {
-            // UNIXTODO: ThreadPool
-            throw new NotImplementedException();
+            if (Interlocked.Exchange(ref _unregisterCalled, 1) == 0)
```

Comment:
Given that for all practical purposes, Unregister is going to be called only once, and the first thing we do is to take a lock, this interlocked operation is unnecessary. Could make it a bool, check it inside the lock, and if true return false.